internal/pkg/service/common/etcdop/op/atomic_test.go

```diff
@@ -13,10 +14,290 @@ import (
 	etcd "go.etcd.io/etcd/client/v3"
 
 	"github.com/keboola/keboola-as-code/internal/pkg/service/common/etcdop"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/common/etcdop/iterator"
 	"github.com/keboola/keboola-as-code/internal/pkg/service/common/etcdop/op"
 	"github.com/keboola/keboola-as-code/internal/pkg/utils/etcdhelper"
+	"github.com/keboola/keboola-as-code/internal/pkg/utils/etcdlogger"
 )
 
+type atomicOpTestCase struct {
+	Name               string
+	Prepare            func(t *testing.T, client etcd.KV) []op.Op
+	ReadPhase          func(t *testing.T, client etcd.KV) []op.Op
+	BreakingChange     func(t *testing.T, client etcd.KV) []op.Op
+	ExpectedWritePhase string
+}
+
+func TestAtomicOp(t *testing.T) {
+	t.Parallel()
+
+	nop := func(_ *testing.T, client etcd.KV) []op.Op {
+		return nil
+	}
+	getKey := func(_ *testing.T, client etcd.KV) []op.Op {
+		return []op.Op{etcdop.Key("key/1").Get(client)}
+	}
+	getPrefix := func(_ *testing.T, client etcd.KV) []op.Op {
+		return []op.Op{etcdop.Prefix("key").GetAll(client).ForEach(func(value *op.KeyValue, header *iterator.Header) error {
+			return nil
+		})}
+	}
+	putKey := func(_ *testing.T, client etcd.KV) []op.Op {
+		return []op.Op{etcdop.Key("key/1").Put(client, "value")}
+	}
+	putTwoKeys := func(_ *testing.T, client etcd.KV) []op.Op {
+		return []op.Op{
+			etcdop.Key("key/1").Put(client, "value"),
+			etcdop.Key("key/2").Put(client, "value"),
+		}
+	}
+	deleteKey := func(_ *testing.T, client etcd.KV) []op.Op {
+		return []op.Op{etcdop.Key("key/1").Delete(client)}
+	}
+
+	cases := []atomicOpTestCase{
+		{
+			// Read Phase gets a key, it doesn't exist, but before the Write Phase, it is created.
+			Name:           "GetKey_CreateKey",
+			Prepare:        nop,
+			ReadPhase:      getKey,
+			BreakingChange: putKey,
+			ExpectedWritePhase: `
+➡️  TXN
+  ➡️  IF:
+  001 "key/1" MOD EQUAL 0
+  ➡️  THEN:
+  001 ➡️  PUT "foo"
+
+✔️  TXN | succeeded: false
```

Comment:
TXN snapshot of the Write Phase:

`key/1` has been read by the get operation in the Read Phase - but it doesn't exist.
So in the Write Phase it must still not exist.